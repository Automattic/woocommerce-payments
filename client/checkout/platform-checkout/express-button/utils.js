/* global wcpayWooPayExpressParams */

/**
 * Retrieves a configuration value.
 *
 * @param {string} name The name of the config parameter.
 * @return {*}         The value of the parameter of null.
 */
export const getWooPayExpressData = ( name ) => {
	if ( 'undefined' === typeof wcpayWooPayExpressParams ) {
		return undefined;
	}

	return wcpayWooPayExpressParams?.[ name ];
};
