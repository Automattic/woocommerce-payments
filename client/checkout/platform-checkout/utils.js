// Waits for the element to exist as in the Blocks checkout, sometimes the field is not immediately available.
export const getTargetElement = ( selector ) => {
	if ( ! selector ) {
		return null;
	}
	return new Promise( ( resolve ) => {
		if ( document.querySelector( selector ) ) {
			return resolve( document.querySelector( selector ) );
		}

		const checkoutBlock = document.querySelector(
			'[data-block-name="woocommerce/checkout"]'
		);

		if ( ! checkoutBlock ) {
			return resolve( null );
		}

		const observer = new MutationObserver( ( mutationList, obs ) => {
			if ( document.querySelector( selector ) ) {
				resolve( document.querySelector( selector ) );
				obs.disconnect();
			}
		} );

		observer.observe( checkoutBlock, {
			childList: true,
			subtree: true,
		} );
	} );
};

export const validateEmail = ( value ) => {
	/* Borrowed from WooCommerce checkout.js with a slight tweak to add `{2,}` to the end and make the TLD at least 2 characters. */
	/* eslint-disable */
	const pattern = new RegExp(
		/^([a-z\d!#$%&'*+\-\/=?^_`{|}~\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]+(\.[a-z\d!#$%&'*+\-\/=?^_`{|}~\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]+)*|"((([ \t]*\r\n)?[ \t]+)?([\x01-\x08\x0b\x0c\x0e-\x1f\x7f\x21\x23-\x5b\x5d-\x7e\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]|\\[\x01-\x09\x0b\x0c\x0d-\x7f\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]))*(([ \t]*\r\n)?[ \t]+)?")@(([a-z\d\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]|[a-z\d\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF][a-z\d\-._~\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]*[a-z\d\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF])\.)+([a-z\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]|[a-z\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF][a-z\d\-._~\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]*[0-9a-z\u00A0-\uD7FF\uF900-\uFDCF\uFDF0-\uFFEF]){2,}\.?$/i
	);
	/* eslint-enable */
	return pattern.test( value );
};
