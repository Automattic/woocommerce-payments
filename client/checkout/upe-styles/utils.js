/**
 * External dependencies
 */
import tinycolor from 'tinycolor2';

/**
 * Generates hover colors from a background color and a text color.
 *
 * @param {string}  backgroundColor Background color, Any format accepted by tinyColor library
 * @param {string}  color Text color, any format accepted by tinyColor library
 * @return {Object} Object with new background color and text color.
 */
export const generateHoverColors = ( backgroundColor, color ) => {
	const hoverColors = {
		backgroundColor,
		color,
	};

	const tinyBackgroundColor = tinycolor( backgroundColor );
	const tinyTextColor = tinycolor( color );

	// If colors are not valid return empty strings.
	if ( ! tinyBackgroundColor.isValid() || ! tinyTextColor.isValid() ) {
		return {
			backgroundColor: '',
			color: '',
		};
	}

	// Darken if brightness > 50 (Storefront Button 51 ), else lighten
	const newBackgroundColor =
		tinyBackgroundColor.getBrightness() > 50
			? tinycolor( tinyBackgroundColor ).darken( 7 )
			: tinycolor( tinyBackgroundColor ).lighten( 7 );

	// Returns provided color if readable, otherwise black or white.
	const mostReadableColor = tinycolor.mostReadable(
		newBackgroundColor,
		[ tinyTextColor ],
		{ includeFallbackColors: true }
	);

	hoverColors.backgroundColor = newBackgroundColor.toRgbString();
	hoverColors.color = mostReadableColor.toRgbString();

	return hoverColors;
};

/**
 * Generates hover rules for UPE using a set of appearance rules as a basis.
 *
 * @param {Object}  baseRules UPE appearance rules to use as a base to generate hover colors
 * @return {Object} Object with generated hover rules.
 */
export const generateHoverRules = ( baseRules ) => {
	const hoverRules = Object.assign( {}, baseRules );

	// If there are no colors, return the same rules as we can not generate hover colors.
	if ( ! baseRules.backgroundColor || ! baseRules.color ) {
		return baseRules;
	}

	const hoverColors = generateHoverColors(
		baseRules.backgroundColor,
		baseRules.color
	);

	hoverRules.backgroundColor = hoverColors.backgroundColor;
	hoverRules.color = hoverColors.color;

	return hoverRules;
};

/**
 * Generates outline style for UPE using outline width, style and color.
 * UPE does not accept the individual properties, we need to concat them.
 *
 * @param {string}  outlineWidth Outline width from computed styles.
 * @param {string}  outlineStyle Outline width from computed styles.
 * @param {string}  outlineColor Outline width from computed styles.
 * @return {string} Object with generated hover rules.
 */
export const generateOutlineStyle = (
	outlineWidth,
	outlineStyle = 'solid',
	outlineColor
) => {
	return outlineWidth && outlineColor
		? [ outlineWidth, outlineStyle, outlineColor ].join( ' ' )
		: '';
};

/**
 * Converts CSS property from dashed format to camel case.
 *
 * @param {string} string CSS property.
 * @return {string} Camel case string.
 */
export const dashedToCamelCase = ( string ) => {
	return string.replace( /-([a-z])/g, function ( g ) {
		return g[ 1 ].toUpperCase();
	} );
};

/**
 * Converts rgba to rgb format, since Stripe Appearances API does not accept rgba format for background.
 *
 * @param {string} color CSS color value.
 * @return {string} Accepted CSS color value.
 */
export const maybeConvertRGBAtoRGB = ( color ) => {
	if ( color.match( /^rgba/ ) ) {
		const colorParts = color.match(
			/^rgba?\((\d+),\s*(\d+),\s*(\d+)(?:,\s*(\d+(?:\.\d+)?))?\)$/
		);
		color = `rgb(${ colorParts[ 1 ] }, ${ colorParts[ 2 ] }, ${ colorParts[ 3 ] })`;
	}
	return color;
};

/**
 * Searches through array of CSS selectors and returns first visible background color.
 *
 * @param {Array} selectors List of CSS selectors to check.
 * @return {string} CSS color value.
 */
export const getBackgroundColor = ( selectors ) => {
	const defaultColor = '#ffffff';
	let color = null;
	let i = 0;
	while ( ! color && i < selectors.length ) {
		const bgColor = window.getComputedStyle(
			document.querySelector( selectors[ i ] )
		).backgroundColor;
		if ( bgColor.match( /^rgba/ ) ) {
			const colorParts = bgColor.match(
				/^rgba?\((\d+),\s*(\d+),\s*(\d+)(?:,\s*(\d+(?:\.\d+)?))?\)$/
			);
			// Check if color is transparent.
			if ( parseInt( colorParts[ 5 ], 10 ) > 0 ) {
				color = bgColor;
			}
		} else {
			color = bgColor;
		}
		i++;
	}
	return color || defaultColor;
};

/**
 * Determines whether background color is light or dark.
 *
 * @param {string} color CSS color value.
 * @return {boolean} True, if background is light; false, if background is dark.
 */
export const isColorLight = ( color ) => {
	let r, g, b;
	if ( color.match( /^rgb/ ) ) {
		color = color.match(
			/^rgba?\((\d+),\s*(\d+),\s*(\d+)(?:,\s*(\d+(?:\.\d+)?))?\)$/
		);
		[ r, g, b ] = color.slice( 1 );
	} else {
		// Convert hex format to rgb
		color = +(
			'0x' + color.slice( 1 ).replace( color.length < 5 && /./g, '$&$&' )
		);

		// eslint-disable-next-line no-bitwise
		r = color >> 16;
		// eslint-disable-next-line no-bitwise
		g = ( color >> 8 ) & 255;
		// eslint-disable-next-line no-bitwise
		b = color & 255;
	}

	// HSP equation from http://alienryderflex.com/hsp.html
	const hsp = Math.sqrt(
		0.299 * ( r * r ) + 0.587 * ( g * g ) + 0.114 * ( b * b )
	);

	// Using the HSP value, determine whether the color is light or dark
	return hsp > 127.5;
};
