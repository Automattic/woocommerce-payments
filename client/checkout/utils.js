/* global wcpay_config, wc */

/**
 * Retrieves a configuration value.
 *
 * @param {string} name The name of the config parameter.
 * @returns {*}         The value of the parameter of null.
 */
export const getConfig = ( name ) => {
	// eslint-disable-next-line camelcase
	const config = ( 'undefined' !== typeof wcpay_config )
		// eslint-disable-next-line camelcase
		? wcpay_config // Classic checkout
		: wc.wcSettings.getSetting( 'woocommerce_payments_data' ); // Blocks

	return config[ name ] || null;
};
