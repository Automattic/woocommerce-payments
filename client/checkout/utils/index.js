/**
 * Normalizes the amount to the accuracy of the minor unit.
 *
 * @param {integer} amount The amount to normalize
 * @param {integer} minorUnit The number of decimal places amount currently represents
 * @param {integer} accuracy The number of decimal places to normalize to
 * @return {integer} The normalized amount
 */
export const normalizeCurrencyToMinorUnit = (
	amount,
	minorUnit = 2,
	accuracy = 2
) => {
	return parseInt( amount * Math.pow( 10, accuracy - minorUnit ), 10 );
};

export const getAppearanceType = () => {
	if ( document.querySelector( '.wp-block-woocommerce-checkout' ) ) {
		return 'blocks_checkout';
	}
};
