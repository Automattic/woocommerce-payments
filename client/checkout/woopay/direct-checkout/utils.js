/**
 * Wait for a given number of milliseconds.
 *
 * @param { number } ms The number of milliseconds to wait.
 * @return {Promise<unknown>} A promise that resolves after the given number of milliseconds.
 */
export const waitMilliseconds = ( ms ) => {
	return new Promise( ( resolve ) => {
		setTimeout( resolve, ms );
	} );
};
