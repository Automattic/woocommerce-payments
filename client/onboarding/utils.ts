/**
 * External dependencies
 */
import { set, toPairs } from 'lodash';
import apiFetch from '@wordpress/api-fetch';

/**
 * Internal dependencies
 */
import { NAMESPACE } from 'data/constants';
import { ListItem } from 'components/grouped-select-control';
import businessTypeDescriptionStrings from './translations/descriptions';
import { Country, OnboardingFields } from './types';

export const fromDotNotation = (
	record: Record< string, unknown >
): Record< string, unknown > =>
	toPairs( record ).reduce( ( result, [ key, value ] ) => {
		return value != null ? set( result, key, value ) : result;
	}, {} );

export const getBusinessTypes = (): Country[] => {
	const data = wcpaySettings?.onboardingFieldsData?.business_types;

	return (
		( data || [] )
			.map( ( country ) => ( {
				...country,
				types: country.types.map( ( type ) => ( {
					...type,
					description: businessTypeDescriptionStrings[ country.key ]
						? businessTypeDescriptionStrings[ country.key ][
								type.key
						  ]
						: businessTypeDescriptionStrings.generic[ type.key ],
				} ) ),
			} ) )
			.sort( ( a, b ) => a.name.localeCompare( b.name ) ) || []
	);
};

export const getMccsFlatList = (): ListItem[] => {
	const data = wcpaySettings?.onboardingFieldsData?.mccs_display_tree;

	// Right now we support only two levels (top-level groups and items in those groups).
	// For safety, we will discard anything else like top-level items or sub-groups.
	const normalizedData = ( data || [] ).filter( ( group ) => {
		if ( ! group?.items ) {
			return false;
		}

		const groupItems =
			group.items?.filter( ( item ) => ! item?.items ) || [];

		return groupItems.length;
	} );

	return normalizedData.reduce( ( acc, group ): ListItem[] => {
		const groupItems =
			group.items?.map(
				( item ): ListItem => {
					return {
						key: item.id,
						name: item.title,
						group: group.id,
						context: item?.keywords
							? item.keywords.join( ' ' )
							: '',
					};
				}
			) || [];

		return [
			...acc,
			{
				key: group.id,
				name: group.title,
				items: groupItems.map( ( item ) => item.key ),
			},
			...groupItems,
		];
	}, [] as ListItem[] );
};

export const persistFlowState = (
	currentStep: string,
	data: OnboardingFields
): Promise< void > =>
	apiFetch( {
		path: `${ NAMESPACE }/onboarding/flow-state`,
		method: 'POST',
		data: { current_step: currentStep, data },
		parse: false,
	} );

export const clearFlowState = (): Promise< void > =>
	apiFetch( {
		path: `${ NAMESPACE }/onboarding/flow-state`,
		method: 'DELETE',
		parse: false,
	} );
