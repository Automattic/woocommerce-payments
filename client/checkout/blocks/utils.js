/**
 * External dependencies
 */
import { useDispatch, useSelect } from '@wordpress/data';

/**
 * Internal dependencies
 */
import { WC_STORE_CART } from 'wcpay/checkout/constants';

/**
 *
 * Custom React hook that provides customer data and related functions for managing customer information.
 * The hook retrieves customer data from the WC_STORE_CART selector and dispatches actions to modify billing and shipping addresses.
 *
 * @return {Object} An object containing customer data and functions for managing customer information.
 */
export const useCustomerData = () => {
	const customerData = useSelect( ( select ) =>
		select( WC_STORE_CART ).getCustomerData()
	);
	const {
		setShippingAddress,
		setBillingData,
		setBillingAddress,
	} = useDispatch( WC_STORE_CART );

	return {
		// Backward compatibility billingData/billingAddress
		billingAddress: customerData.billingAddress || customerData.billingData,
		// Backward compatibility setBillingData/setBillingAddress
		setBillingAddress: setBillingAddress || setBillingData,
		setShippingAddress,
	};
};
