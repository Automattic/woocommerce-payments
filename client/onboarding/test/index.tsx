/**
 * External dependencies
 */
import React from 'react';
import { render } from '@testing-library/react';
import { mocked } from 'ts-jest/utils';

/**
 * Internal dependencies
 */
import OnboardingPage from '..';
import AddBusinessInfoTask from '../tasks/add-business-info-task';
import SetupCompleteTask from '../tasks/setup-complete-task';

jest.mock( '../tasks/add-business-info-task', () => jest.fn() );
jest.mock( '../tasks/setup-complete-task', () => jest.fn() );

describe( 'OnboardingPage', () => {
	beforeEach( () => {
		mocked( AddBusinessInfoTask ).mockReturnValue(
			<p>Add Business Info</p>
		);
		mocked( SetupCompleteTask ).mockReturnValue( <p>Setup Complete</p> );
	} );

	afterEach( () => {
		jest.restoreAllMocks();
	} );

	it( 'renders page', () => {
		const { container: list } = render( <OnboardingPage /> );
		expect( list ).toMatchSnapshot();
	} );
} );
