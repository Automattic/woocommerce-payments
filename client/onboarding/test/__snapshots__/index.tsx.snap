// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OnboardingPage renders page 1`] = `
<div>
  <div
    class="components-card is-size-large wcpay-onboarding-card wcpay-homescreen-card css-1xs3c37-CardUI e1q7k77g0"
  >
    <div
      class="components-card__body is-size-large css-xmjzce-BodyUI e1q7k77g3"
    >
      <div>
        <ul>
          <p>
            Add Business Info
          </p>
          <p>
            Setup Complete
          </p>
        </ul>
      </div>
    </div>
  </div>
</div>
`;
